src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * The backoff time is between the minimum and maximum backoff time, based on the number of attempts.
+ * An exponential backoff strategy is used, with a jitter factor to prevent clients from retrying at the same time.
+ *
+ * The backoff time is calculated as follows:
+ *  - `basic backoff time` = `MinimumBackoffInMs` * 2 ^ `attempts`, and it is no larger than the `MaximumBackoffInMs`.
+ *  - based on jitter ratio, the jittered time is between [-1, 1) * `JitterRatio` * basic backoff time.
+ *  - the final backoff time is the basic backoff time plus the jittered time.
+ *
+ * Note: the backoff time usually is no larger than the refresh interval, which is specified by the user.
+ *  - If the interval is less than the minimum backoff, the interval is used.
+ *  - If the interval is between the minimum and maximum backoff, the interval is used as the maximum backoff.
+ *  - Because of the jitter, the maximum backoff time is actually `MaximumBackoffInMs` * (1 + `JitterRatio`).
+ */
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxAttempts = 63;
+const JitterRatio = 0.25;
+
+export class RefreshTimer {
+    private _minBackoff: number;
+    private _maxBackoff: number;
+    private _attempts: number;
+    private _backoffEnd: number; // Timestamp
+    constructor(
+        private _interval: number
+    ) {
+        this._minBackoff = Math.min(this._interval, MinimumBackoffInMs);
+        this._maxBackoff = Math.min(this._interval, MaximumBackoffInMs);
+        this._attempts = 0;

```

Comment:
In python provider, _attempts is initialized as 1.

Reply:
AFAIU, the `MaxAttempt` (copied from dotnet provider) should be the guardian of overflow for 64 bit integer. 
Here Number.MAX_SAFE_INTEGER = 2^53-1, so in JS we should use 53. 
And the name is confusing. I would use `MaxSafeExponential` instead.

Reply:
fixed as above in e0c9736

Reply:
Discussed with Yan offline. `_attempts` will be renamed to `_failedAttempts`

Reply:
> In python provider, _attempts is initialized as 1.  Also checked the dotnet provider
> It seems like the python provider is wrong.

In .NET, we increase the number of attempts before calculating backoff: https://github.com/Azure/AppConfiguration-DotnetProvider/blob/e9a3aa5b49958bd6190fb91c7a16d10ac46e3967/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs#L84C1-L85

In Python, we calculate the backoff and then increase the attempts: https://github.com/Azure/azure-sdk-for-python/blob/98274a8e4684d868471654cfe92469d452316428/sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py#L404-L406

Both are correct, just doing things differently.

Reply:
Python defaults to 1 attempt, not 0, so we increment after. My logic or it is that if you ever request the attempt number it is always the attempt you are on.

Reply:
https://github.com/Azure/azure-sdk-for-python/blob/98274a8e4684d868471654cfe92469d452316428/sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py#L421

For python, if the initial value is 1, above calculated time is 2x min duration, which is not consistent with dotnet and JS in this PR. Might change it to attempts-1?

Reply:
The first time we calculate backoff, attempts should be 1. This is consistent in .NET and Python. Am I missing something?

Reply:
https://github.com/Azure/AppConfiguration-DotnetProvider/blob/e9a3aa5b49958bd6190fb91c7a16d10ac46e3967/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/Extensions/TimeSpanExtensions.cs#L68

If I understand correctly, in dotnet the attempts is 1, and the calculated time is min duration. In python, the attempts is also 1, but the calculated time is 2 * min duration.

Reply:
Ah yes, you're right. I forgot about that extra check in .NET when attempts=1. Python should also do that @mrm9084